service/relayer/src/cache.rs

```diff
@@ -0,0 +1,171 @@
+            update(value);
+            &*value
+        })
+    }
+
+    pub async fn fetch<E>(
+        &mut self,
+        fetch_fresh: impl AsyncFnOnce() -> Result<T, E>,
+        maximum_age: Duration,
+    ) -> Result<&T, E> {
+        self.fetch_update(fetch_fresh, maximum_age, |_| {}).await
+    }
+
+    pub async fn fetch_update<E>(
+        &mut self,
+        fetch_fresh: impl AsyncFnOnce() -> Result<T, E>,
+        maximum_age: Duration,
+        and_transform: impl FnOnce(&mut T),
+    ) -> Result<&T, E> {
+        if self.value.is_some()
+            && self
+                .updated_at
+                .elapsed()
+                .is_ok_and(|elapsed| elapsed <= maximum_age)
+        {
+            #[allow(clippy::unwrap_used, reason = "Guaranteed by .is_some() call")]
+            let v = self.value.as_mut().unwrap();
+            and_transform(v);
+            return Ok(v);
+        }
+
+        let mut v = fetch_fresh().await?;
+        self.updated_at = SystemTime::now();
+        and_transform(&mut v);
+        let r = self.value.insert(v);
+        Ok(r)
+    }
+}
+
+#[derive(Debug)]
+pub struct Cache {
+    request: mpsc::Sender<CacheRequest>,
+}
+
+impl Cache {
+    pub fn new(near: Near, gas_price_refresh: Duration, nonce_refresh: Duration) -> Self {
+        let (send, mut recv) = mpsc::channel::<CacheRequest>(64);
+
+        tokio::spawn(async move {
+            let mut gas_price = CacheRecord::empty();
+            let mut nonce = HashMap::<(AccountId, PublicKey), CacheRecord<u64>>::new();
+            let block_hash = Arc::new(RwLock::new(CryptoHash::new()));
+
+            let update_gas = || async { near.fetch_gas_price().await };
+            let update_nonce = |(account_id, public_key)| {
+                || async {
+                    let (nonce, hash) = near.fetch_nonce(account_id, public_key).await?;
+                    *block_hash.write().await = hash;
+                    Ok::<_, JsonRpcError<RpcQueryError>>(nonce + 1)
+                }
+            };
+
+            while let Some(request) = recv.recv().await {
+                match request {
+                    CacheRequest::GasPrice(sender) => {
+                        let fresh = gas_price.fetch(update_gas, gas_price_refresh).await;
+                        #[allow(clippy::unwrap_used)]
+                        if let Ok(price) = fresh {
+                            sender.send(*price).unwrap();
+                        } else if let Some(price) = gas_price.stale() {
+                            tracing::warn!("Failed to fetch gas price, sending stale value.");
+                            sender.send(*price).unwrap();
+                        } else {
+                            // We should only ever *not* have a stale value on startup, so this should be a "fail-fast" case.
+                            tracing::error!("Failed to fetch gas price, no stale value cached.");
+                            panic!("Failed to fetch gas price.");
```

Comment:
This will panic the spawned task, but we should also shut down the relayer itself if this happens, so I'd also pass in a `watch` channel for communicating with the main task that this failed, thus we can gracefully shut down anything else, or at the very least have better error messaging